lxd/networks.go

```diff
@@ -1898,3 +1930,79 @@ func (n *network) Update(newNetwork api.NetworkPut) error {
 
 	return nil
 }
+
+func (n *network) spawnForkDNS(listenAddress string) error {
+	// Get the list of nodes
+	nodes, err := cluster.List(n.state)
+	if err != nil {
+		logger.Errorf("Failed to start forkdns for network '%s': %v", n.name, err)
+		return err
+	}
+
+	localAddress, err := node.HTTPSAddress(n.state.Node)
+	if err != nil {
+		logger.Errorf("Failed to start forkdns for network '%s': %v", n.name, err)
+		return err
+	}
+
+	// Grab the network address from the various nodes
+	addresses := []string{listenAddress}
+
+	cert := n.state.Endpoints.NetworkCert()
+	for _, node := range nodes {
+		address := strings.TrimPrefix(node.URL, "https://")
+		if address == localAddress {
+			continue
+		}
+
+	again:
+		client, err := cluster.Connect(address, cert, false)
+		if err != nil {
+			time.Sleep(30 * time.Second)
+			goto again
+		}
+
+		state, err := client.GetNetworkState(n.name)
+		if err != nil {
+			time.Sleep(30 * time.Second)
+			goto again
+		}
+
+		for _, addr := range state.Addresses {
+			if addr.Family != "inet" || addr.Scope != "global" {
+				continue
+			}
+
+			addresses = append(addresses, addr.Address)
+			break
+		}
+	}
+
+	// Setup the dnsmasq domain
+	dnsDomain := n.config["dns.domain"]
+	if dnsDomain == "" {
+		dnsDomain = "lxd"
+	}
+
+	// Spawn the daemon
+	cmd := exec.Cmd{}
+	cmd.Path = n.state.OS.ExecPath
+	cmd.Args = []string{n.state.OS.ExecPath, "forkdns", fmt.Sprintf("%s:1053", listenAddress), dnsDomain}
+	cmd.Args = append(cmd.Args, addresses...)
+
+	err = cmd.Start()
+	if err != nil {
+		logger.Errorf("Failed to start forkdns for network '%s': %v", n.name, err)
+		return err
+	}
+
+	// Write the PID file
+	pidPath := shared.VarPath("networks", n.name, "forkdns.pid")
+	err = ioutil.WriteFile(pidPath, []byte(fmt.Sprintf("%d\n", cmd.Process.Pid)), 0600)
```

Comment:
If this fails you don't really have a way of killing the daemon later. Shouldn't you kill the  `cmd.Start()` process or are you doing it somewhere else?

Reply:
fixed